src/aosm/azext_aosm/tests/latest/unit_test/test_processors/test_core_arm_processor.py

```diff
@@ -0,0 +1,175 @@
+        artifact_details = self.processor.get_artifact_details()
+        mock_arm_template_path = os.path.join(mock_vnf_directory, "ubuntu-template.json")
+        mock_artifact = [LocalFileACRArtifact(
+            artifact_name="test-artifact-name",
+            artifact_type="ArmTemplate",
+            artifact_version="1.1.1",
+            file_path=mock_arm_template_path,
+        )]
+
+        # Testing each individial part of artifact are equal
+        self.assertEqual(artifact_details[0][0].artifact_name, mock_artifact[0].artifact_name)
+        self.assertEqual(artifact_details[0][0].artifact_version, mock_artifact[0].artifact_version)
+        self.assertEqual(artifact_details[0][0].artifact_type, mock_artifact[0].artifact_type)
+        self.assertEqual(artifact_details[0][0].file_path, mock_artifact[0].file_path)
+        # Ensure no list of LocalFileBuilders is returned, as this is only for NSDs
+        self.assertEqual(artifact_details[1], [])
+        assert True
+
+    def test_generate_nf_application(self):
+        """Test generate nf application for Azure Core ARM Processor."""
+        # Check type is correct, other functionality is tested in appropriate functions
+        # (such as test_generate_artifact_profile)
+        nf_application = self.processor.generate_nf_application()
+        self.assertIsInstance(nf_application, AzureCoreNetworkFunctionArmTemplateApplication)
+
+    def test_generate_resource_element_template(self):
+        """Test generate RET for Azure Core ARM Processor"""
+        result = self.processor.generate_resource_element_template()
+
+        # Assert the expected output
+        expected_template = ArmResourceDefinitionResourceElementTemplateDetails(
+            name="test-name",
+            depends_on_profile=DependsOnProfile(
+                install_depends_on=[], uninstall_depends_on=[], update_depends_on=[]
+            ),
+            configuration=ArmResourceDefinitionResourceElementTemplate(
+                template_type=TemplateType.ARM_TEMPLATE.value,
+                parameter_values=json.dumps({}),
+                artifact_profile=NSDArtifactProfile(
+                    artifact_store_reference=ReferencedResource(id=""),
+                    artifact_name=self.processor.input_artifact.artifact_name,
+                    artifact_version=self.processor.input_artifact.artifact_version,
+                ),
+            ),
+        )
+        # Assert each parameter equal
+        self.assertEqual(result.name, expected_template.name)
+        self.assertEqual(result.depends_on_profile, expected_template.depends_on_profile)
+        self.assertEqual(result.configuration.artifact_profile,
+                         expected_template.configuration.artifact_profile)
+
+    def test_generate_parameters_file(self):
+        """ Test generate parameters file for Azure Core ARM Processor"""
+
+        # Mock private function
```

Comment:
What's the rationale for mocking `_generate_mapping_rule_profile()`?

Possible reasons I can think of:
1. actually calling the method is too slow / requires too much setup / requires some real system that isn't practical to use in a unit test
2. 'expected_params' is a really important input to this test, and setting it by the inputs to `_generate_mapping_rule_profile()` is too cumbersome / not flexible enough
3. You want to isolate this test from any bugs in `_generate_mapping_rule_profile()` so it's easier to locate the bug if a test fails.

I don't think #1 applies here because there's a unit test for this method. #3 also sounds fine - if there's a serious problem with `_generate_mapping_rule_profile()`, both tests will fail, and the developer should look at the narrower-scope test first when debugging?

#2 could potentially be an issue, although I think the argument for it would be stronger if you wanted to test `generate_parameters_file()` with lots of different params rather than just one. Worth noting though that needing to mock a private method for this reason can be a code smell - if `_generate_mapping_rule_profile()` is doing something so complicated that it needs to be mocked out to make the tests simpler, perhaps it should be refactored out into a separate class?

Personally I try to only use mocking where absolutely required (because it tends to make tests harder to understand, slower and more brittle (if the interface of `_generate_mapping_rule_profile()` changes, these mocking lines need to be updated), and would try to avoid it in cases like this. Maybe you don't need to _test_ the mapping rule profile in this test case because it's covered by another one, then you could just not assert anything about it?

Reply:
We chatted :)